internal/model/tempnameutil.go

```diff
@@ -0,0 +1,34 @@
+// Copyright (C) 2014 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package model
+
+import (
+	"crypto/md5"
+	"fmt"
+	"path/filepath"
+	"strings"
+)
+
+type tempNamer struct {
+	prefix string
+}
+
+func (t tempNamer) IsTemporary(name string) bool {
+	return strings.HasPrefix(filepath.Base(name), t.prefix)
+}
+
+func (t tempNamer) TempName(name string) string {
+	tdir := filepath.Dir(name)
+	tbase := filepath.Base(name)
+	if len(tbase) > 100 {
+		hash := md5.New()
+		hash.Write([]byte(name))
+		tbase = fmt.Sprintf("%s%s", tbase[0:100], hash.Sum(nil))
```

Comment:
Actually, I think this needs either more work or less. If the filename is a minor Russian novel, as in the original issue, the `tbase[0:100]` will more likely than not cut off the last codepoint somewhere in the middle, creating an invalid UTF-8 string. So we either need to treat the string as runes and iterate over it to get a whole number of code points, or just say fuck it and skip the name completely once it's too long. I'd suggest just letting the name through up to 240 characters (as we add 15), and once the length exceeds that just use the hash and nothing else.